src/java/org/apache/cassandra/service/accord/serializers/DepsSerializers.java

```diff
@@ -74,164 +83,294 @@ public DepsSerializer(UnversionedSerializer<Range> tokenRange)
-                Range[] ranges = new Range[rangeCount];
-                for (int i = 0; i < rangeCount; i++)
-                    ranges[i] = tokenRange.deserialize(in);
-
-                int txnIdCount = in.readUnsignedVInt32();
-                TxnId[] txnIds = new TxnId[txnIdCount];
-                for (int i = 0; i < txnIdCount; i++)
-                    txnIds[i] = CommandSerializers.txnId.deserialize(in);
-
-                int rangesToTxnIdsCount = in.readUnsignedVInt32();
-                int[] rangesToTxnIds = new int[rangesToTxnIdsCount];
-                for (int i = 0; i < rangesToTxnIdsCount; i++)
-                    rangesToTxnIds[i] = in.readUnsignedVInt32();
-
-                rangeDeps = RangeDeps.SerializerSupport.create(ranges, txnIds, rangesToTxnIds);
+                Range[] ranges = KeySerializers.rangeArray.deserialize(in);
+                TxnId[] txnIds = CommandSerializers.txnId.deserializeArray(TxnId[]::new, in);
+                int[] txnIdsToRanges = null, rangesToTxnIds = null;
+                if (0 != (flags & RANGES_BY_TXNID)) txnIdsToRanges = deserializePackedXtoY(txnIds.length, ranges.length, in);
+                else rangesToTxnIds = deserializePackedXtoY(ranges.length, txnIds.length, in);
+                rangeDeps = RangeDeps.SerializerSupport.create(ranges, txnIds, rangesToTxnIds, txnIdsToRanges);
             }
             return deserialize(keyDeps, rangeDeps, in);
         }
 
+        private static int[] deserializePackedXtoY(int xCount, int yCount, DataInputPlus in) throws IOException
+        {
+            int length = in.readUnsignedVInt32();
+            int[] xtoy = new int[length];
+
+            if ((xCount <= 1 || yCount <= 1) && (xtoy.length == xCount + yCount || xCount == 0 || yCount == 0))
+            {
+                // no point serializing as can be directly inferred
+                if (xCount == 1)
+                {
+                    xtoy[0] = xtoy.length;
+                    for (int i = 0 ; i < yCount ; ++i)
+                        xtoy[1 + i] = i;
+                }
+                else if (yCount == 1)
+                {
+                    for (int i = 0 ; i < xCount ; ++i)
+                        xtoy[i] = xCount + i + 1;
+                }
+                else if (yCount == 0)
+                {
+                    for (int i = 0 ; i < xCount ; ++i)
+                        xtoy[i] = xCount;
+                }
+                else
+                {
+                    Invariants.require(length == 0);
+                }
+            }
+            else
+            {
+                deserializePackedInts(xtoy, 0, xCount, xtoy.length, in);
+                deserializePackedInts(xtoy, xCount, xtoy.length, yCount - 1, in);
+            }
+            return xtoy;
+        }
+
+        private static void deserializePackedInts(int[] vs, int from, int to, int max, DataInputPlus in) throws IOException
```

Comment:
I think it's worth describing the bit packing scheme here in a 4-5 sentence comment.